### PR TITLE
Problem: Discoverable assets & decentralisation.

### PR DESCRIPTION
**Pull request by ldmberman**
Implement a contract for registering decentralised identifiers, according to OEP-7.

There is [a pending update of OEP-7](https://github.com/oceanprotocol/OEPs/pull/77/files). Also, a storage optimisation is going to be included there and here:

From @aaitor :
> there are 2 main approaches to implement this:
>1. mapping between assetDid => providerUrl
>2. mapping between assetDid => providerDid => providerUrl
>the first approach is simpler but requires more bytes on-chain. Every asset would require to have >associated a url (a lot of bytes)
>
>the second one is cheaper, all the assets associated with the same provider are related with the same >providerDID (sorter than a URL so cheaper), but the resolving is bit more complex

### CODE REVIEW

File: contracts/didregistry/DIDRegistry.sol [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+contract DIDRegistry is Ownable {
+    enum DocumentType { Asset }
+
+    struct Identity {
+        address owner;
+        string did;
+        DocumentType _type;
+    }
+
+    // FIXME if we index the string field (did), there gonna be problems reading it
```
Comment (H34D) [inline contracts/didregistry/DIDRegistry.sol]: We fully migrated to web3 1.0 in thise repo and squid-js. This is no issue anymore.
See https://github.com/oceanprotocol/keeper-contracts/pull/123 for details.

Comment (ldmberman) [reply contracts/didregistry/DIDRegistry.sol]: My tests reproduce the issue in our current setup, with `web3 = 1.0.0-beta.36`.

Comment (H34D) [reply contracts/didregistry/DIDRegistry.sol]: hum

**Comment by ldmberman** [comment]
Regarding the provider URL optimisation, the current implementation is flexible in that this is how documents can be registered:

```
const registry = await DIDRegistry.new()

// register a provider with an URL
registry.registerAttribute(providerDID, providerType, providerURL, 'http://...')

// register an asset with a provider DID
registry.registerAttribute(assetDID, assetType, providerDID, 'did:ocn:...')
```

**Comment by aaitor** [comment]
@H34D , @ldmberman , @billbsing , what do you think of having the DID as a bytes32 type instead of string? 
PROS:
* Cheaper
* Easier to search on it
CONS:
* No fragments/paths as part of the did
* Limited to 24 characters: 32 characters - 8 characters ("did:ocn:")

**Comment by H34D** [comment]
> @H34D , @ldmberman , @billbsing , what do you think of having the DID as a bytes32 type instead of string?
> PROS:
> 
> * Cheaper
> * Easier to search on it
> 
> CONS:
> 
> * No fragments/paths as part of the did
> * Limited to 24 characters: 32 characters - 8 characters ("did:ocn:")

lets go, bytes32

File: contracts/didregistry/DIDRegistry.sol (hunk 1 above)
Comment (H34D) [reply contracts/didregistry/DIDRegistry.sol]: `
{
  "tx": "0x3d2831348f0ee443b201a61a20c6e6040b066cff2d9fe28ad494f3115b75a972",
  "receipt": {
    "transactionHash": "0x3d2831348f0ee443b201a61a20c6e6040b066cff2d9fe28ad494f3115b75a972",
    "transactionIndex": 0,
    "blockHash": "0xb80efb26a661d1bedba821e4a545abad986c1a1d2e1a09ecfa74c71561fcd336",
    "blockNumber": 333,
    "gasUsed": 57636,
    "cumulativeGasUsed": 57636,
    "contractAddress": null,
    "logs": [
      {
        "logIndex": 0,
        "transactionIndex": 0,
        "transactionHash": "0x3d2831348f0ee443b201a61a20c6e6040b066cff2d9fe28ad494f3115b75a972",
        "blockHash": "0xb80efb26a661d1bedba821e4a545abad986c1a1d2e1a09ecfa74c71561fcd336",
        "blockNumber": 333,
        "address": "0xA70B85Ef627939c9A8e8C772E2E76C98084caa01",
        "data": "0xa1cfabf665d259ef90a83ee60a7b0a186c19a425e53ce25efb80a866eba6f89e00000000000000000000000000000000000000000000000000000000000000a0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000e0000000000000000000000000000000000000000000000000000000000000014d000000000000000000000000000000000000000000000000000000000000001f6469643a6f636e3a746573742d72656769737465722d617474726962757465000000000000000000000000000000000000000000000000000000000000000012687474703a2f2f6578616d706c652e636f6d0000000000000000000000000000",
        "topics": [
          "0x3f59c0320125ffa479c2ea4c3f03ee266e7641c3abea13a6def51ad85dc91a30",
          "0x000000000000000000000000b21f531c831471eb3515183232a66db229070f47",
          "0x70726f7669646572000000000000000000000000000000000000000000000000"
        ],
        "type": "mined",
        "id": "log_f6773ff5"
      }
    ],
    "status": true,
    "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000008000000000000000000002000000000000000000000000000000000000000000000000000000040000000000210000000000000000000000000000000000000000000000000000000200000000000000000000000002000000000000000000000000000000000000000000000001000008000000000400000000000000000000000000000000000000000000000040000000000000000000000000000000000000000000000000000000000000000000000000000000000000000"
  },
  "logs": [
    {
      "logIndex": 0,
      "transactionIndex": 0,
      "transactionHash": "0x3d2831348f0ee443b201a61a20c6e6040b066cff2d9fe28ad494f3115b75a972",
      "blockHash": "0xb80efb26a661d1bedba821e4a545abad986c1a1d2e1a09ecfa74c71561fcd336",
      "blockNumber": 333,
      "address": "0xA70B85Ef627939c9A8e8C772E2E76C98084caa01",
      "type": "mined",
      "id": "log_f6773ff5",
      "event": "DIDAttributeRegistered",
      "args": {
        "0": "0xa1cfabf665d259ef90a83ee60a7b0a186c19a425e53ce25efb80a866eba6f89e",
        "1": "did:ocn:test-register-attribute",
        "2": "0xB21F531c831471Eb3515183232a66db229070F47",
        "3": "0",
        "4": "0x70726f7669646572000000000000000000000000000000000000000000000000",
        "5": "http://example.com",
        "6": "14d",
        "__length__": 7,
        "didHash": "0xa1cfabf665d259ef90a83ee60a7b0a186c19a425e53ce25efb80a866eba6f89e",
        "did": "did:ocn:test-register-attribute",
        "owner": "0xB21F531c831471Eb3515183232a66db229070F47",
        "_type": "0",
        "key": "0x70726f7669646572000000000000000000000000000000000000000000000000",
        "value": "http://example.com",
        "updatedAt": "14d"
      }
    }
  ]
}

`

Comment (H34D) [reply contracts/didregistry/DIDRegistry.sol]: that contains:

`
        "didHash": "0xa1cfabf665d259ef90a83ee60a7b0a186c19a425e53ce25efb80a866eba6f89e",
        "did": "did:ocn:test-register-attribute",
        "owner": "0xB21F531c831471Eb3515183232a66db229070F47",
        "_type": "0",
        "key": "0x70726f7669646572000000000000000000000000000000000000000000000000",
        "value": "http://example.com",
`

Comment (ldmberman) [reply contracts/didregistry/DIDRegistry.sol]: Weird.. Thankfully, we do not have this problem either way since we have decided to use `bytes32` there.

**Pull request merged**